Add reforge optimizer setting for Hit/Expertise preference

Saji-Saji opened the pull request:
This pull request introduces a new option to the `ReforgeOptimizer` that allows caster classes to prefer Hit Rating over Expertise Rating when suggesting reforges. This helps ensure that, when both stats are available as reforging options, Hit is prioritized, which is more relevant for caster classes. The option is enabled by default for all Mage specializations. Additionally, this seems to make reforging decisions faster when enabled for casters as it removes possible options.

Key changes:

**Reforge logic improvements:**

* Added a new `preferHitOverExpertise` option to `ReforgeOptimizerOptions` and implemented logic in `ReforgeOptimizer` to filter out Expertise reforges in favor of Hit reforges for items that don't natively have either stat. [[1]](diffhunk://#diff-3906d7c7cc0efedbac6c68c2c2b52b6e47ad40f6e17ead6fb3b5b6074d91b4e9R85-R87) [[2]](diffhunk://#diff-3906d7c7cc0efedbac6c68c2c2b52b6e47ad40f6e17ead6fb3b5b6074d91b4e9R225) [[3]](diffhunk://#diff-3906d7c7cc0efedbac6c68c2c2b52b6e47ad40f6e17ead6fb3b5b6074d91b4e9R272) [[4]](diffhunk://#diff-3906d7c7cc0efedbac6c68c2c2b52b6e47ad40f6e17ead6fb3b5b6074d91b4e9R1330-R1354)

**Mage specialization updates:**

* Enabled the new `preferHitOverExpertise` option by default for Arcane, Fire, and Frost Mage simulation UIs to improve reforge suggestions for caster gear. [[1]](diffhunk://#diff-f55ce04f26dac79926e56a893fa12886f65a650ac7eed47ffcba2b180ac5d610R175) [[2]](diffhunk://#diff-387723de2e5e8ff81a9842512baa74723b192c4dd2aba5124a549ecb22dade36R312) [[3]](diffhunk://#diff-0adc18f65184ecea596501cc7eb1eb4111370ec41a8fda176f3b3b8c35ccf848R191)

Saji-Saji committed:
Add reforge optimizer setting for preferring a reforge to hit instead of to expertise when both options are present. Enable this by default for all mage specs

1337LutZ commented:
Can we not bake this into the reforger itself instead of making it a boolean? Just checking for all caster specs
```
isCaster := character.Class == proto.Class_ClassMage ||
				character.Class == proto.Class_ClassWarlock ||
				character.Class == proto.Class_ClassPriest ||
				character.Spec == proto.Spec_SpecBalanceDruid ||
				character.Spec == proto.Spec_SpecRestorationDruid ||
				character.Spec == proto.Spec_SpecElementalShaman ||
				character.Spec == proto.Spec_SpecRestorationShaman ||
				character.Spec == proto.Spec_SpecHolyPaladin ||
				character.Spec == proto.Spec_SpecMistweaverMonk
```
(but then in JS)

Saji-Saji committed:
Remove preferHitOverExpertise option, auto-detect caster specs

- Removed preferHitOverExpertise boolean parameter from ReforgeOptimizer
- Replaced with automatic caster spec detection
- Casters (Mage, Warlock, Priest, Balance/Resto Druid, Ele/Resto Shaman, Holy Paladin, Mistweaver Monk) now automatically prefer Hit over Expertise
- Removed preferHitOverExpertise: true from all mage spec UIs

1337LutZ commented:
> > Can we not bake this into the reforger itself instead of making it a boolean? Just checking for all caster specs
> > ```
> > isCaster := character.Class == proto.Class_ClassMage ||
> > 				character.Class == proto.Class_ClassWarlock ||
> > 				character.Class == proto.Class_ClassPriest ||
> > 				character.Spec == proto.Spec_SpecBalanceDruid ||
> > 				character.Spec == proto.Spec_SpecRestorationDruid ||
> > 				character.Spec == proto.Spec_SpecElementalShaman ||
> > 				character.Spec == proto.Spec_SpecRestorationShaman ||
> > 				character.Spec == proto.Spec_SpecHolyPaladin ||
> > 				character.Spec == proto.Spec_SpecMistweaverMonk
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > (but then in JS)
> 
> Baked in for all caster specs, Hesitant if we should do this for Holy Paladin or Mistweaver, as they may want to actually use expertise at times? We don't exactly support it, but someone may be using it.

Yea I think you can leave those out indeed. As they can also proc RoRo since they are considered Melee. (need to change that still)
